students/etwum/lesson09/mailroom_part5.py

```diff
@@ -0,0 +1,213 @@
+            print('{:<25} $ {:<20}{:^14} $ {:<15}'.format(*x.values()))
+
+    def send_letter_all(self):
+        print('\n----------------------------------------------')
+        # creates a letter to every donor by writing them to separate text files
+        for x in self.donor_list:
+
+            # opens or creates a new text file for writing based on the donor name
+            with open('{file_name}.txt'.format(file_name=x.donor_data()["donor name"]), 'w') as f1:
+
+                # writes a thank you letter to the text file
+                f1.write('Dear {donor_name},\n\nThank you for your generous donation(s) of ${donations:.2f} '
+                         'to our charity.'.format(donor_name=x.donor_data()["donor name"],
+                                                  donations=x.donor_data()["total donations"]))
+                print('Letter sent to  {a}'.format(a=x.donor_data()["donor name"]))
+
+        print('----------------------------------------------\n')
+
+
+donor1 = DonorData({"donor name": "Lionel Messi", "total donations": 1000000.00, "number donations": 5,
+                     "avg donation": 20000.00})
+donor2 = DonorData({"donor name": "Thierry Henry", "total donations": 500, "number donations": 1,
+                     "avg donation": 500})
+donor3 = DonorData({"donor name": "Michael Jordan", "total donations": 45000, "number donations": 3,
+                     "avg donation": 15000})
+donor4 = DonorData({"donor name": "Kobe Bryant", "total donations": 8000, "number donations": 2,
+                     "avg donation": 4000})
+
+
+run = Interaction([donor1,donor2,donor3,donor4])
+
+user_selection = {1: run.send_thank_you, 2: run.create_report, 3: run.send_letter_all}
+
+
+def options():
+    # function for returning a user selection
+    return user_selection
+
+
+def main():
+    print("Welcome to the Charity Mail Room")
+    print("------------------------------------------------------------------------")
+    str_choice = 0
+    # ****Input/Output****
+    while str_choice != 4:
+        # Option menu
+        print("""
+            Menu of Options
+            1) Send Thank You Note
+            2) Create Report
+            3) Send Letter to Everyone
+            4) Exit Program
+            """)
+
+        # try/except block to make sure the user inputs a valid option
+        try:
+            str_choice = int(input("Which option would you like to perform? Input a number [1 to 4] "))
+            if str_choice in user_selection:
+                # returns a value from user_selection dictionary in the options function
+                # then runs the function from the dictionary
+                options()[str_choice]()
+            elif str_choice == 4:
+                print("\nExiting Program")
+            else:
+                raise Exception
+        except Exception:
+            print("Please input a valid option: 1, 2, 3, or 4")
```

Comment:
you could simplify this a bit:
```
try:
    str_choice = input("")  # you can just use strings for `user_selection` -> {"1": run.send_thank_you}
    user_selection[str_choice]()
except KeyError:
    print("Please enter valid options")
```
you could also use `while True` and have '4' be redirected to `sys.exit` function